README.md
# FluPipeline

`FluPipeline` is a command line program for processing Influenza sequencing data. It takes as input a folder of paired-end read files, finds the best flu strain to find variants against, and outputs a variant calls, mapping and coverage data, and taxonomic placement.

## Overview

<img src="docs/overview.png" alt="FluPipeline overview" width=1000>

**Highlighted Features**

- Detailed VCF tables in CSV format for several different variant callers (bcftools, bbmap, lofreq, freebayes)
- Call intra host variants
- Automatic reference strain selection
- Custom reference strains can be added
- Automatic thread selection selection to run samples in parallel
- Downsample reads and run pipeline
- Simple installation
- Automated testing

---

## Installation

Enter `FluPipeline` directory and install/load environment.

```sh
cd /path/to/FluPipeline

conda env create --file /install/environment.yml

conda activate testenv
```

## Test installation

Verify that expected outputs are created by `FluPipeline`.

```sh
cd /path/to/FluPipeline

python -m unittest tests.integration.test_detect_variants -v

python -m unittest tests.integration.test_best_reference -v

python FluPipeline.py --runtest
```

## Example set up

Simple skeleton code to run`FluPipeline` on a set of reads:

```sh
cd /path/to/FluPipeline

python FluPipeline.py \
--base_directory /path/to/output \
--sequence_directory /path/to/reads \
--threads 8 \
--force_base_directory \
--intrahost_variant_caller bbtools
```

## detect_variants


`FluPipeline` is built around `detect_variants`. You can run `detect_variants` as you would any other program and supply it a read pair, an output directory, and a reference fasta.  Check out all the options:

```sh
python -m bin.detect_variants -h
```

Simple skeleton code to run `detect_variants` on a read pair:

```sh
cd /path/to/FluPipeline

python -m bin.detect_variants \
--baseDir /path/to/output/directory \
--logDir /path/to/output/directory \
--variant_caller bbtools \
--refGenomeFasta /path/to/reference/genome.fasta \
--R1 /path/to/R1.fastq.gz \
--R2 /path/to/R2.fastq.gz \
--consensus_sequence
```

If you want to compare different settings/variant callers, one short cut is to take an existing sample out from FluPipeline and run the intra-host variant calling step with different parameters:

```sh
cd /path/to/FluPipeline

python -m bin.detect_variants \
--build_input_from /path/to/sampleOutputs/sample \
--output_name my_output_folder \
--variant_caller lofreq \
--BWAmappingScore 10 \
--minVariantPhredScore 5 \
--minimum_read_depth 1
```

## Monitor run

In a different terminal, check how many samples have finished/are running/are errored out. Sometimes samples with poor sequencing depth will error out, but this will not stop FluPipeline from processing other samples.

```sh
cd /path/to/FluPipeline

python bin/check_finished.py /path/to/base/directory
```


# Options

FluPipeline has many options, check them out below:

```sh
cd /path/to/FluPipeline

python FluPipeline.py -h
```

```md
usage: FluPipeline v0.7.0 [-h] [--base_directory] [--reference_directory] [--sequence_directory] [--use_fasta] [--force] [--force_base_directory] [--threads] [--max_mem_per_thread] [--strain_sample_depth] [--downsample] [--use_strain]
                          [--base_quality] [--keep_duplicates] [--remove_NTs_from_alignment_ends] [--keep_trimmed_reads] [--min_read_mapping_score] [--gap_open_penalty] [--gap_extension_penalty] [--major_variant_caller]
                          [--intrahost_variant_caller] [--single_pass] [--min_variant_phred_score] [--min_variant_frequency] [--major_variant_frequency] [--major_indel_frequency] [--minimum_read_depth] [--runtest]

options:
  -h, --help            show this help message and exit
  --base_directory      directory that run will output data to [./FluPipeline_output]
  --reference_directory
                        directory containing reference strain files (.gb or .fasta (see --use_fasta flag if you want to use fasta files)) [script_path/references]
  --sequence_directory
                        directory containing fastq sequence files (.gz format) [None]
  --use_fasta           fasta format: fasta file(s) contain all eight segments sequences. All segments must have a single name (only letters, numbers, and 3 underscores. At the end of the name there should be an underscore followed by the
                        segment number. Example: an_example_1. [False]
  --force               overwrite existing sample files. [False]
  --force_base_directory
                        overwrite existing directory. [False
  --threads             number of samples to process in paralell. one sample is one read pair [4]
  --max_mem_per_thread
                        automatically determines the number of threads to use based on memory per thread supplied (in Gb) [None]
  --strain_sample_depth
                        number of random reads to use to determine strain assignment. [2000]
  --downsample          downsample all read files to these many reads. [-1 (no downsampling)]
  --use_strain          name of strain. No file extensions in name. [None]
  --base_quality        keep reads that have at least an average of this phred-scaled value. [30]
  --keep_duplicates     do not conduct read deduplication. [False]
  --remove_NTs_from_alignment_ends
                        remove this many bases from the left and right of each read prior to mapping. [3]
  --keep_trimmed_reads  keep trimmed reads used for analysis. [False]
  --min_read_mapping_score
                        keep reads that mapped above or equal to this MAPQ value. [10]
  --gap_open_penalty    The gap open penalty used by BWA during alignment. [6]
  --gap_extension_penalty
                        The gap extension penalty used by BWA during alignment. [1]
  --major_variant_caller
                        variant caller to use (bcftools, bbmap, lofreq, freebayes). [bcftools]
  --intrahost_variant_caller
                        intra host variant caller to use (bcftools, bbmap, lofreq, freebayes). [lofreq]
  --single_pass         only call variants one time (no variants from consensus sequence will be called). [False]
  --min_variant_phred_score
                        keep all variants above or equal to this phred-scaled value. [20]
  --min_variant_frequency
                        keep all variants with allele frequencies above or equal this value. [0.05]
  --major_variant_frequency
                        keep all major variants with allele frequencies above or equal this value. [0.5]
  --major_indel_frequency
                        keep all major indels with allele frequencies above or equal this value. [0.8]
  --minimum_read_depth
                        Mask/ignore all bases and variants at or below this read depth. [10]
  --runtest             Run an in silico test to make sure FluPipeline is working correctly. [False]
  ```

# Citation

[FluPipeline - Alex McFarland - github.com/agmcfarland/FluPipeline](https://github.com/agmcfarland/FluPipeline)

[Follow me on twitter! @alexmcfarland_](https://twitter.com/alexmcfarland_)



docs/changelog.md

# 0.7.0

- Removed `consensus_masking_threshold` option.

- Removed `keep_all_intermediate_files` option.

- Removed `masked_nextclade` option.

- Removed `masked_ivar` option.

- Removed `testbiin` option

- Added unittests for modules `bin.best_reference`, `bin.detect_variants`.

- `runtest` now outputs to the default temporary directory.

- `runtest` now has `compare_TestResults` to validate the observed results with known results.

- Added `gap_open_penalty` option.

- Added `gap_extension_penalty` option.


# 0.6.0

- Added picard to remove duplicates at the alignment level

- Fixed bug where files used for processing in best_reference.py were not removed after script was finished.

- Removed sampleReports directory

- Options using masked consensus sequences are availble but will still used unmasked consensensus sequence. This option may be updated or removed in the future.

# 0.5.0

- Major rewrite to remove R dependency.


FluPipeline.py
#!/usr/bin/env python

import sys
import os 
from os.path import join as pjoin
import multiprocessing as mp
from multiprocessing import Pool
import pandas as pd
import re
import glob
import shutil
import subprocess
from datetime import datetime
from Bio.Seq import Seq
from Bio import SeqIO, SeqFeature
import argparse
import logging
import psutil
import tempfile

global script_path
script_path = os.path.split(os.path.realpath(os.path.abspath(__file__)))[0]
sys.path.append(script_path)
from pylibs.processing_functions import *
from pylibs.processing_classes import SequencingSample, RunLogger
from pylibs.pipeline import flu_Pipeline
from pylibs.insilico_test import create_TestData, compare_TestResults
from pylibs.helper_functions import automatic_ThreadUsage, call_Command
from _version import __version__
pd.set_option('display.max_columns', None)

def run_FluPipeline(args):
	'''
	Runs through all steps of FluPipeline
	'''
	start_run_timer = datetime.now()
	## prepare working directory
	if args.sequence_directory == pjoin(script_path,'references'): # prevent the user from using the default reference directory in combination with the --use_fasta flag
		args.use_fasta = False

	if args.force_base_directory == True: 	# remove base directory if it exiss
		try:
			shutil.rmtree(args.base_directory)
		except:
			pass

	if args.use_strain == 'None':
		args.use_strain = None

	## set up main directories
	os.makedirs(args.base_directory, exist_ok=True)
	os.chdir(args.base_directory)
	os.makedirs(pjoin(args.base_directory,'sampleLogs'), exist_ok=True)
	os.makedirs(pjoin(args.base_directory,'sampleOutputs'), exist_ok=True)

	## start up run logger
	run_logger = RunLogger(directory=os.getcwd(),filename='runLog')
	run_logger.initialize_FileHandler()
	run_logger.add_StreamHandler()
	run_logger.logger.info('Starting FluPipeline...\n')
	run_logger.logger.info(f'Version {__version__}\n')

	## log computing resources available and requested
	run_logger.logger.info('Total memory available: {}Gb\n'.format(round(psutil.virtual_memory()[0]/(1024.0**3)),2))
	run_logger.logger.info('Threads available: {}\n'.format(os.cpu_count()))
	if args.max_mem_per_thread != None: # check whether automatic thread number was requested and log calculations
		run_logger.logger.info('Automatic thread number selection\n')
		run_logger.logger.info('Threads use: {}\n'.format(args.threads))
		run_logger.logger.info('Estimated memory per thread: {}Gb\n'.format(args.max_mem_per_thread))
		run_logger.logger.info('Estimated max memory usage: {}Gb\n'.format(args.threads*args.max_mem_per_thread))

	## log argument inputs, reference strains, and samples	
	run_logger.logger.info('Program arguments:\n')
	arguments_list = vars(args)
	for k,v in arguments_list.items():
		run_logger.logger.info('{}: {}\n'.format(k,v))

	run_logger.logger.info('Reference strains:\n')
	if args.use_fasta == True:
		[run_logger.logger.info('{}\n'.format(os.path.basename(g))) for g in glob.glob(pjoin(args.reference_directory,'*.fasta'))]
	else:
		[run_logger.logger.info('{}\n'.format(os.path.basename(g))) for g in glob.glob(pjoin(args.reference_directory,'*.gb'))]

	run_logger.logger.info('Samples:\n') # contains sample name, fastq R1 file, fastq R2 file
	for s in glob.glob(pjoin(args.sequence_directory,'*_R1_*.fastq.gz')):
		sample = SequencingSample()
		sample.get_DataFromReadPairs(read1_filename=s)
		run_logger.logger.info('{}\n'.format(sample.samplename))


	## start data processing------------------------------------------------------------

	## collect all samples to process in a list for multiprocessing submission
	sample_submission = []
	for f in glob.glob(pjoin(args.sequence_directory,'*_R1_*.fastq.gz')):
		# create sample object with necessary data and check that it exists
		sample = SequencingSample()
		sample.get_DataFromReadPairs(read1_filename=f)
		sample.get_SampleDirPath(directory=args.base_directory)
		sample.check_ReadExistence(directory=args.sequence_directory)

		# append sample run parameters to master run list
		sample_submission.append([
			args.base_directory, #baseDirectory
			script_path, #softwareDir
			sample, #sample
			args.sequence_directory, #sequenceDataDir
			args.reference_directory, #referenceStrainsDir
			args.force, #force_overwrite
			args.strain_sample_depth, #strain_sample_depth
			args.downsample, #downsample
			args.min_variant_phred_score, #min_variant_phred_score
			args.remove_NTs_from_alignment_ends, #remove_NTs_from_alignment_ends
			args.min_read_mapping_score, #min_read_mapping_score
			args.gap_open_penalty, #gap_open_penalty
			args.gap_extension_penalty, #gap_extension_penalty
			args.base_quality, #base_quality
			args.keep_duplicates, #keep_duplicates
			args.min_variant_frequency, #min_variant_frequency
			args.use_strain, #use_strain
			args.keep_trimmed_reads, #keep_trimmed_reads
			args.major_variant_frequency, #major_variant_frequency
			args.major_indel_frequency, #major_indel_frequency
			args.minimum_read_depth, #minimum_read_depth
			args.major_variant_caller, #major_variant_caller
			args.intrahost_variant_caller, #intrahost_variant_caller
			args.single_pass #single_pass
			])

	run_logger.logger.info('Total samples to process: {}\n'.format(len(sample_submission))) # total samples ran

	if len(sample_submission) == 0:
		run_logger.logger.exception('No samples available to process. Check directory for correctly formatted fastq file names.\n')

	if args.use_fasta == True:
		run_logger.logger.info('Using fasta files for reference\n')
	else:
		run_logger.logger.info('Using gbk files for reference\n')
		# convert genbank files to uniformally-formatted fasta. see convert_GBKTOFasta for details.
		gbk_reference_files = glob.glob(pjoin(args.reference_directory,'*.gb'))
		[convert_GBKToFasta(filename=f.replace('.gb','')) for f in gbk_reference_files]


	run_logger.logger.info('Processing samples...\n')
	## run listed samples through flu_Pipeline function
	with Pool(processes=args.threads) as p:
		p.starmap(flu_Pipeline, sample_submission)

	run_logger.logger.info('Finished processing samples\n')

	## end data processing--------------------------------------------------------------


	## run reports
	run_logger.logger.info('Creating sample reports and run report...\n')
	os.chdir(script_path)
	call_Command(cmd=
		f'python -m bin.summarize --baseDir {args.base_directory} --sequenceDir {args.sequence_directory}',
		logger_=run_logger,
		shell_=True)
	os.chdir(args.base_directory)

	## write software versions
	run_logger.logger.info('Recording software/package versions...\n')
	call_Command(cmd=
		'conda list --export > softwareVersions.txt',
		logger_=run_logger,
		shell_=True)

	## end run
	run_logger.logger.info('Finished running FluPipeline\n')
	run_logger.logger.info('Total time: {} (H:M:S)\n'.format(datetime.now()-start_run_timer))
	run_logger.logger.info('Run outputs stored in {}\n'.format(args.base_directory))	



def main(args = None):
	'''
	Main function for parsing arguments
	'''
	if args is None:
		args = sys.argv[1:]

	# create parser
	parser = argparse.ArgumentParser(prog = 'FluPipeline v'+ str(__version__))

	# add arguments to parser
	# main arguments
	parser.add_argument('--base_directory',type=str ,default=pjoin(os.getcwd(),'FluPipeline_output'), help='directory that run will output data to [./FluPipeline_output]', metavar='')
	parser.add_argument('--reference_directory',type=str ,default=pjoin(script_path,'references'), help='directory containing reference strain files (.gb or .fasta (see --use_fasta flag)) [script_path/references]', metavar='')
	parser.add_argument('--sequence_directory',type=str ,default=None, help='directory containing fastq sequence files (.gz format) [None]', metavar='')
	parser.add_argument('--use_fasta', action='store_true', default=False, help='fasta format: fasta file(s) contain all eight segments sequences. All segments must have a single name (only letters, numbers, and 3 underscores. At the end of the name there should be an underscore followed by the segment number. Example: an_example_1. [False]')

	# remove files/folders
	parser.add_argument('--force', action='store_true', default=False, help='overwrite existing sample files. [False]')
	parser.add_argument('--force_base_directory', action='store_true', default=False, help='overwrite existing directory. [False')

	# parallel jobs
	parser.add_argument('--threads',type=int, default=4, help='number of samples to process in paralell. one sample is one read pair [4]', metavar='')
	parser.add_argument('--max_mem_per_thread', type=int, default=None, help='automatically determines the number of threads to use based on memory per thread supplied (in Gb) [None]', metavar='')

	# strain detection
	parser.add_argument('--strain_sample_depth', type=int, default=2000, help='number of random reads to use to determine strain assignment. [2000]', metavar='')
	parser.add_argument('--downsample', type=int, default=-1, help='downsample all read files to these many reads. [-1 (no downsampling)]', metavar='')
	parser.add_argument('--use_strain', type=str, default=None, help='name of strain. No file extensions in name. [None]', metavar='')

	# read processing 
	parser.add_argument('--base_quality', type=int, default=30, help='keep reads that have at least an average of this phred-scaled value. [30]', metavar='')
	parser.add_argument('--keep_duplicates',  action='store_true', default=False, help='do not conduct read deduplication.  [False]')
	parser.add_argument('--remove_NTs_from_alignment_ends', type=int, default=3, help='remove this many bases from the left and right of each read prior to mapping. [3]', metavar='')
	parser.add_argument('--keep_trimmed_reads', action='store_true', default=False, help='keep trimmed reads used for analysis.  [False]')
	
	# read mapping
	parser.add_argument('--min_read_mapping_score', type=int, default=10, help='keep reads that mapped above or equal to this MAPQ value. [10]', metavar='')
	parser.add_argument('--gap_open_penalty', type=int, default=6, help='The gap open penalty used by BWA during alignment. [6]', metavar='')
	parser.add_argument('--gap_extension_penalty', type=int, default=1, help='The gap extension penalty used by BWA during alignment. [1]', metavar='')
	
	# variant caller
	parser.add_argument('--major_variant_caller', type=str, default='bcftools', choices=['bcftools','lofreq','bbtools', 'freebayes'], help='variant caller to use (bcftools, bbmap, lofreq, freebayes). [bcftools]', metavar='')
	parser.add_argument('--intrahost_variant_caller', type=str, default='bbtools', choices=['bcftools','lofreq','bbtools', 'freebayes'], help='intra host variant caller to use (bcftools, bbmap, lofreq, freebayes). [bbtools]', metavar='')
	parser.add_argument('--single_pass', action='store_true', default=False, help='only call variants one time (no variants from consensus sequence will be called).  [False]')
	

	# variant calling parameters
	parser.add_argument('--min_variant_phred_score', type=int, default=20, help='keep all variants above or equal to this phred-scaled value. [20]', metavar='')
	parser.add_argument('--min_variant_frequency', type=float, default=0.05, help='keep all variants with allele frequencies above or equal this value. [0.05]', metavar='')
	parser.add_argument('--major_variant_frequency', type=float, default=0.5, help='keep all major variants with allele frequencies above or equal this value. [0.5]', metavar='')
	parser.add_argument('--major_indel_frequency', type=float, default=0.8, help='keep all major indels with allele frequencies above or equal this value. [0.8]', metavar='')
	parser.add_argument('--minimum_read_depth', type=int, default=10, help='Mask/ignore all bases and variants at or below this read depth. [10]', metavar='')

	# run test
	parser.add_argument('--runtest', action='store_true', default=False, help='Run an in silico test to make sure FluPipeline is working correctly. [False]')

	# create args opbject with arguments
	args = parser.parse_args()

	## calculate threads to use if --max_mem_per_thread was supplied
	if args.max_mem_per_thread != None:
		calculated_threads_to_use = automatic_ThreadUsage(max_mem_per_thread = args.max_mem_per_thread)
		args.threads = calculated_threads_to_use #overwrite args.threads with the new value

	## create test data if --runtest argument was supplied
	if args.runtest == True:
		testDir =  tempfile.mkdtemp(dir = tempfile.gettempdir())
		print(f'Outputting test to: {testDir}')
		args.base_directory = pjoin(testDir,'output')
		args.sequence_directory = pjoin(testDir,'data')
		args.reference_directory = pjoin(script_path,'references')
		args.force == True
		args.force_base_directory = True
		args.keep_trimmed_reads = True
		args.single_pass = True
		create_TestData(testDir=testDir, referenceStrainsDir=args.reference_directory)
		run_FluPipeline(args=args)
		compare_TestResults(existing_test_data_dir = pjoin(script_path, 'tests', 'data'), new_test_data_dir = pjoin(testDir, 'output'))
		sys.exit()	

	##---Run FluPipline ---##	
	run_FluPipeline(args=args)

if __name__ == '__main__':
	sys.exit(main())

